### DISCUSSION TITLE
My C# Console script doesn't finish when using(Py.GIL()) finish it's execution

### DISCUSSION THREAD

**Discussion by Maskliv** [pythonnet/pythonnet#1687] [Q&A]
Here is my code 

```
using System;
using System.Collections.Generic;
using Python.Runtime;

namespace PythonCsharp

{
    class Program

    {
        static void Main(string[] args)

        {

            Console.WriteLine("Python code begins");

            // Set Runtime.PythonDLL
            Runtime.PythonDLL = @"C:\Users\JAC\AppData\Local\Programs\Python\Python37\python37.dll";

            // Ejecución desde python
            using (Py.GIL())
            {
                //Example of opencv

                dynamic cv = Py.Import("cv2");

                dynamic img = cv.imread(@"C:\Users\JAC\Desktop\ProyectosCsharp\PythonCsharp\imagenes\lenna.png");

                dynamic imgBW = cv.cvtColor(img, cv.COLOR_RGB2GRAY);

                cv.imshow("Lenna.png", img);

                cv.imshow("LennaBW", imgBW);

                cv.waitKey(0);

                cv.destroyAllWindows();
            }

            Console.WriteLine("Python code ends");

        } // -------> Here my code gets stuck and doesn't finish it's execution
    }
}
```
Code works fine but doesn't finish the execution and when i debug, can see that on the line marked the execution gets stuck , when i comment the "using(Py.GIL())" the code finish normally as a regular console script write on C#.

![image](https://user-images.githubusercontent.com/68347324/137211553-652b9821-8be6-4884-b396-ecdc78ff104f.png)

**Comment by OlegZuev** [comment]
Same problem

**Comment by OlegZuev** [comment]
@Maskliv Found solution. You have to call PythonEngine.Shutdown(); after you finished working with Python.

**Comment by MrM40** [comment]
If I call `PythonEngine.Shutdown()` after `using (Py.GIL()) {..}` I get an "Segmentation fault" error?
Debian 11, Bullseye
Python 3.9.2

UPDATE:
It only happens when using `numpy`. But I remember now it says somewhere there is issues with this particular module and it relate to a bug in the module. Don't see the issue in another module (`apt`, the Debian package management tool).
For me it look like everything is not released/disposed when Py.GIL() is disposed, either using a "using" block or calling Py.GIL.Dispose() manually.
But it's not because numpy is "hanging" outside pythonnet. If I restart pytonnet (restart my C# app") `numpy` is not hanging anymore, so it's surprising  pythonnet cannot release it fully.

**Comment by bitshop** [comment]
Calling shutdown using stripped down numpy sample I get:

**Fatal Python error: auto-releasing thread-state, but no thread-state for this thread**

```csharp
using System;
using Python.Runtime;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello, World!");
            using (Python.Runtime.Py.GIL())
            {
                dynamic np = Python.Runtime.Py.Import("numpy");
                Console.WriteLine(np.cos(np.pi * 2));
                PythonEngine.Shutdown();
            }
        }
    }
}

```

**Comment by MrM40** [comment]
I would expect to place `PythonEngine.Shutdown();` _after_ the using statement.

**Comment by CodingMadness** [comment]
@MrM40  doing as you suggest, works also for me properly without any issue! 

I would tend to close such issues actually.
